### PR TITLE
feat: add initial markdown provider

### PR DESCRIPTION
**Pull request by adamkdean**
Issue: https://github.com/dadi/web/issues/96
Usage: https://github.com/adamkdean/dadi-web-md

Is this the kind of functionality you'd be interested in, @abovebored? (See usage above)

image

Note: this a WIP, please don't merge yet @jimlambie

### CODE REVIEW

**Comment by adamkdean** [comment]
I've added search, filter, sort, count, and field methods:

image

```
{
  "datasource": {
    "key": "posts",
    "name": "A statically loaded greeting.",
    "source": {
      "type": "markdown",
      "path": "workspace/posts"
    },
    "count": 10,
    "sort": {
      "date": -1
    },
    "filter": {
      "published": true
    }
  }
}
```

But...I'm seeing some inconsistencies in how we implement things like sorting etc. Going to spend some time and have a look through and try and tidy it all up.

**Comment by adamkdean** [comment]
I'm just in the progress of giving the providers something of an audit. Some of them have inconsistent syntax for things like fields etc, and sometimes the features aren't implemented at all.

Part of this will be pagination for markdown.

Will update you soon!

**Comment by adamkdean** [comment]
@abovebored haven't forgotten about this one. Will get MD provider released first and tidy up others when possible. Doesn't make sense to hold up the MD provider.

**Comment by adamkdean** [comment]
@abovebored pagination added as requested. Sorry for the delay! 🛌

**Comment by abovebored** [comment]
FYI, I added `metadata` to allow more complex pagination (as API would add to mongo datasets) and renamed the frontmatter `metadata` variable to `attributes` to avoid any confusion between the two.

File: dadi/lib/providers/markdown.js [hunk 1]
```diff
@@ -143,7 +145,7 @@ MarkdownProvider.prototype.parseRawDataAsync = function parseRawDataAsync (data,
   callback(posts)
 }
 
-// From @dadi/api
+// Stolen from @dadi/api
```
Comment (adamkdean) [inline dadi/lib/providers/markdown.js]: 😂

**Comment by adamkdean** [comment]
Good idea with the pagination metadata, @abovebored.

**Comment by adamkdean** [comment]
@jimlambie I would be happy to write docs for this. I also want to fix up the getting started doc as well.

File: dadi/lib/providers/markdown.js (hunk 1 above)
Comment (abovebored) [inline dadi/lib/providers/markdown.js]: Always with dem jokes 😬

**Comment by jimlambie** [comment]
Could you please reopen the PR against branch `feature/version-2.0`? Thanks!

**Comment by adamkdean** [comment]
@abovebored would you be happy to handle the docs for this, on second thought? 

https://media.giphy.com/media/3osxY5DBiplbJ47i4o/giphy.gif

**Comment by abovebored** [comment]
@adamkdean http://i0.kym-cdn.com/photos/images/newsfeed/001/210/371/1bb.gif

**Pull request merged**